### PR TITLE
Implementation of the eFlux Concept

### PR DESCRIPTION
**Pull request by tpfau**
This PR adds an implementation of the eFlux concept as presented in 
```
    Interpreting Expression Data with Metabolic Flux Models: Predicting Mycobacterium tuberculosis Mycolic Acid Production
    Colijn C, Brandes A, Zucker J, Lun DS, Weiner B, et al. (2009)
    PLOS Computational Biology 5(8): e1000489. https://doi.org/10.1371/journal.pcbi.1000489
```

**I hereby confirm that I have:**

- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [X] Selected `develop` as a target branch (top left drop-down menu)

*(Note: You may replace [ ] with [X] to check the box)*

### CODE REVIEW

File: src/dataIntegration/transcriptomics/EFlux/eFlux.m [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+%                             * noiseCount - number of noisy controls to create if noise is tested (default: 10)
+%                             * noiseFun - The noise function to use, has to be a function handle taking 2 arguments (mean and std)
+%                             * noiseStd - The standard deviation(s) to use. Either a single value (used for all values, or a vector with length equal to controlExpression.value).
+%                             * controlData - a struct (like controlExpression which has a value matrix with multiple values per controlExpression to determine the noise distribution. If provided with testNoise == false, the values from this struct will be used to determine the noise.
+%                             * minSum:           Switch for the processing of Genetic data. If false, ORs in the GPR will be treated as min. If true(default), ORs will be treated as addition.
+%                             * softBounds:       Whether to use soft bounds for the infered constraints or to add flexibility variables (default: false).
+%                             * weightFactor:     The weight factor for soft bounds (default: 1) 
+% OUTPUT:
```
Comment (laurentheirendt) [inline src/dataIntegration/transcriptomics/EFlux/eFlux.m]: `OUTPUTS`

File: src/dataIntegration/transcriptomics/EFlux/eFlux.m [hunk 2]
```diff
@@ -0,0 +1,142 @@
+function [foldChange,standardError,solControl,solCondition] = eFlux(model,controlExpression,conditionExpression,varargin)
```
Comment (laurentheirendt) [inline src/dataIntegration/transcriptomics/EFlux/eFlux.m]: May you name the subfolder `eFlux` and add more comments in the different functions?

File: src/dataIntegration/transcriptomics/EFlux/eFlux.m [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+% remove enforcing bounds (contradict eFlux concept)
+if(any(model.lb > 0 | model.ub < 0))
+    warning('Enforcing bounds for the following fluxes have been removed:\n%s', strjoin(model.rxns((model.lb > 0 | model.ub < 0)),'\n'));
+    model.lb(model.lb > 0) = 0;
+    model.ub(model.ub < 0) = 0;
+end
+    
+if testNoise
+    noisyControl = repmat(columnVector(controlExpression.value),1,noiseCount);
+    if isempty(controlData) && isempty(noiseStd)
```
Comment (laurentheirendt) [inline src/dataIntegration/transcriptomics/EFlux/eFlux.m]: The check of `isempty(controlData)` that throws an error shall be placed before `if testNoise`. Then, the repetitive `if ~isempty(controlData)` conditions  can be omitted.

File: src/dataIntegration/transcriptomics/EFlux/verifyEFluxExpressionStruct.m [hunk 4]
```diff
@@ -0,0 +1,26 @@
+function tf = verifyEFluxExpressionStruct(model,expression)
```
Comment (laurentheirendt) [inline src/dataIntegration/transcriptomics/EFlux/verifyEFluxExpressionStruct.m]: May you add a few more comments here?

File: src/dataIntegration/transcriptomics/preprocessing/mapExpressionToReactions.m [hunk 5]
```diff
@@ -12,17 +12,24 @@
 %                               format as model.genes
 %       .value                  Vector containing corresponding expression
 %                               value (FPKM/RPKM)
+% OPTIONAL INPUTS:
```
Comment (laurentheirendt) [inline src/dataIntegration/transcriptomics/preprocessing/mapExpressionToReactions.m]: `OPTIONAL INPUT`

File: src/dataIntegration/transcriptomics/preprocessing/selectGeneFromGPR.m [hunk 6]
```diff
@@ -14,28 +14,50 @@
 %                   (as returned by "findUsedGeneLevels.m")
 %   parsedGPR:      GPR matrix as returned by "GPRparser.m"
 %
+% OPTIONAL INPUTS:
```
Comment (laurentheirendt) [inline src/dataIntegration/transcriptomics/preprocessing/selectGeneFromGPR.m]: `OPTIONAL INPUT`

File: src/reconstruction/refinement/GPRLogic/Node.m [hunk 7]
```diff
@@ -68,19 +68,21 @@
...
         
-        deleteLiteral(self,literalID)
+        tf = deleteLiteral(self,literalID, keepClauses)
         % Delete a literal from this Node and all children
         % USAGE:
         %    newHead = node.deleteLiteral(literalID)
         % 
         % INPUT:
         %    literalID: The LiteralID (As string or number)
+        % OPTIONAL INPUT:
```
Comment (laurentheirendt) [inline src/reconstruction/refinement/GPRLogic/Node.m]: There has to be an empty line here.

File: src/reconstruction/refinement/GPRLogic/Node.m [hunk 8]
```diff
@@ -196,9 +198,50 @@ function removeDuplicateLiterals(self)
                 toRemove = setdiff(literals,toKeep);
                 self.children(toRemove) = [];
             end
+        end               
+        
+        function tf = hasNestedChildren(self)
+        % Check, whether this Node has nested children (i.e. whether it has
```
Comment (laurentheirendt) [inline src/reconstruction/refinement/GPRLogic/Node.m]: These internal functions need proper header formatting.

File: src/reconstruction/refinement/GPRLogic/Node.m [hunk 9]
```diff
@@ -207,31 +250,58 @@ function removeDuplicateLiterals(self)
             %    geneNames:   the genes in the order represented by the
             %                 literals (which represent positions)
             %
+            % OPTIONAL INPUTS:
```
Comment (laurentheirendt) [inline src/reconstruction/refinement/GPRLogic/Node.m]: `OPTIONAL INPUT`

File: src/reconstruction/refinement/GPRLogic/Node.m [hunk 10]
```diff
@@ -207,31 +250,58 @@ function removeDuplicateLiterals(self)
...
+                    childliterals = normNode.children(i).getLiterals();
                     pos = cellfun(@str2num, childliterals);
                     genePos{i} = pos;
                     geneSets{i} = geneNames(sort(pos));
                 end
             end
+        end                
+        
+        function deleteLiterals(self, literals, keepClauses)
+        % Remove all duplicate literal nodes in this node.
```
Comment (laurentheirendt) [inline src/reconstruction/refinement/GPRLogic/Node.m]: same comment here about the internal function

File: test/verifiedTests/dataIntegration/testEFlux/Aerobic.txt [hunk 11]
```diff
@@ -0,0 +1,4298 @@
+^SAMPLE = GSM1126445
```
Comment (laurentheirendt) [inline test/verifiedTests/dataIntegration/testEFlux/Aerobic.txt]: May you fetch this data from somewhere else? What is the source of this test data?

File: src/dataIntegration/transcriptomics/EFlux/eFlux.m (hunk 3 above)
Comment (tpfau) [reply src/dataIntegration/transcriptomics/EFlux/eFlux.m]: I don't really get this. 
missing/empty controlData will only throw an error if testNoise is active. 
I'm not sure how you want this changed.

File: src/reconstruction/refinement/GPRLogic/Node.m (hunk 8 above)
Comment (tpfau) [reply src/reconstruction/refinement/GPRLogic/Node.m]: How so?
Whats wrong with the formatting?

File: test/verifiedTests/dataIntegration/testEFlux/Aerobic.txt (hunk 11 above)
Comment (tpfau) [reply test/verifiedTests/dataIntegration/testEFlux/Aerobic.txt]: I'll update it to download the data from GEO

File: src/dataIntegration/transcriptomics/EFlux/eFlux.m (hunk 3 above)
Comment (laurentheirendt) [reply src/dataIntegration/transcriptomics/EFlux/eFlux.m]: ok

**Comment by laurentheirendt** [comment]
Thanks @tpfau 🌵 👍

**Pull request merged**